corehq/util/couch_doc_processor.py

```diff
@@ -0,0 +1,372 @@
+        iteration_name = "{}/{}".format(iteration_key, " ".join(doc_types))
+        self.iteration_id = hashlib.sha1(iteration_name).hexdigest()
+        self.maintain_retry_state = maintain_retry_state
+        try:
+            self.state = db.get(self.iteration_id)
+        except ResourceNotFound:
+            # new iteration
+            self.state = {
+                "_id": self.iteration_id,
+                "doc_type": "ResumableDocsByTypeIteratorState",
+                "retry": {},
+                "retry_cleared": [],
+
+                # for humans
+                "name": iteration_name,
+                "timestamp": datetime.utcnow().isoformat()
+            }
+            args = {}
+        else:
+            # resume iteration
+            args = self.state.get("offset", {}).copy()
+            if args:
+                assert args.get("startkey"), args
+                doc_type = args["startkey"][0]
+                # skip doc types before offset
+                doc_types = doc_types[doc_types.index(doc_type):]
+            else:
+                # non-retry phase of iteration is complete
+                doc_types = []
+        args.update(
+            view_name='all_docs/by_doc_type',
+            chunk_size=chunk_size,
+            log_handler=ResumableDocsByTypeLogHandler(self),
+            include_docs=True,
+            reduce=False,
+        )
+        self.view_args = args
+        self.doc_types = doc_types
+
+    def __iter__(self):
+        args = self.view_args
+        for doc_type in self.doc_types:
+            if args.get("startkey", [None])[0] != doc_type:
+                args.pop("startkey_docid", None)
+                args["startkey"] = [doc_type]
+            args["endkey"] = [doc_type, {}]
+            for result in paginate_view(self.db, **args):
+                yield result['doc']
+
+        retried = {}
+        while self.state["retry"] != retried:
+            for doc_id, retries in list(self.state["retry"].iteritems()):
+                if doc_id in self.state["retry_cleared"]:
+                    retried[doc_id] = retries
+                    continue
+                if retries == retried.get(doc_id):
+                    continue  # skip already retried
+                retried[doc_id] = retries
+                try:
+                    yield self.db.get(doc_id)
+                except ResourceNotFound:
+                    pass
+
+        # save iteration state without offset to signal completion
+        self.state.pop("offset", None)
+        if self.maintain_retry_state:
+            for doc_id in self.state["retry_cleared"]:
+                try:
+                    del self.state["retry"][doc_id]
+                except KeyError:
+                    pass
+        else:
+            self.state["retry"] = {}
+        self.state["retry_cleared"] = []
+        self._save_state()
+
+    def mark_doc_processed(self, doc):
+        """Mark a document as having been processed successfully
```

Comment:
This method seems to do something that contradicts the docstring of the `retry` method:

> every retry document will be yielded each time the iterator is stopped and resumed during the retry phase.